lib/identity_cache/cache_fetcher.rb

```diff
@@ -24,27 +75,179 @@ def fetch_multi(keys, &block)
+      lock = nil
+      using_fallback_key = false
+      expiration_options = EMPTY_HASH
+      (lock_wait_limit + 2).times do # +2 is for first attempt and retry with fallback key
+        result = fetch_or_take_lock(key, old_lock: lock, **expiration_options)
+        case result
+        when FillLock
+          lock = result
+          if lock.client_id == client_id # have lock
+            data = begin
+              yield
+            rescue
+              @cache_backend.cas(key, **expiration_options) do |value|
+                break unless FillLock.cache_value?(value)
+                lock = FillLock.from_cache(value)
+                lock.mark_failed
+                lock.cache_value
+              end
+              raise
+            end
+
+            unless fill_with_lock(key, data, lock, expiration_options)
+              unless using_fallback_key
+                # fallback to storing data in the fallback key so it is available to clients waiting on the lock
+                expiration_options = fallback_key_expiration_options(fill_lock_duration)
+                @cache_backend.write(lock_fill_fallback_key(key, lock), data, **expiration_options)
+              end
+            end
+            return data
+          else
+            raise LockWaitTimeout if lock_wait_limit <= 0
+            lock_wait_limit -= 1
+
+            # If fill failed in the other client, then it might be failing fast
+            # so avoid waiting the typical amount of time for a lock wait. The
+            # semian gem can be used to handle failing fast when the database is slow.
+            if lock.fill_failed?
+              return fetch_without_fill_lock(key) { yield }
+            end
+
+            # lock wait
+            sleep(fill_lock_duration)
+            # loop around to retry fetch_or_take_lock
+          end
+        when IdentityCache::DELETED # interrupted by cache invalidation
+          if lock && !using_fallback_key
+            # cache invalidated during lock wait, try fallback key
+            using_fallback_key = true
+            key = lock_fill_fallback_key(key, lock)
+            expiration_options = fallback_key_expiration_options(fill_lock_duration)
+            # loop around to retry with fallback key
+          else
+            # cache invalidation prevented lock from being taken
+            return yield
+          end
+        when nil # Errors talking to memcached
+          return yield
+        else # hit
+          return result
+        end
+      end
+      raise "unexpected number of loop iterations"
+    end
+
+    def cas_or_add(key, expiration_options = EMPTY_HASH)
+      yielded = false
+      swapped = @cache_backend.cas(key, expiration_options) do |value|
+        yielded = true
+        yield value
       end
       unless yielded
-        result = yield
-        add(key, result)
+        data = yield nil
+        swapped = add(key, data, expiration_options)
```

Comment:
Technically, the return value here represents `stored` not `swapped`. `updated` is the most accurate term to encompass both the `@cache_backend.cas` and `add` return values.

Reply:
Or, I suppose, `upserted`.